plexus-compilers/plexus-compiler-javac/src/main/java/org/codehaus/plexus/compiler/javac/JavacCompiler.java

```diff
@@ -659,26 +666,44 @@ static List<CompilerMessage> parseModernStream(int exitCode, BufferedReader inpu
+                    } else if (stackTraceLineCount > 0) {
+                        // Extract stack trace from end of buffer
+                        String[] lines = bufferAsString.split("\\R");
+                        int linesTotal = lines.length;
+                        buffer = new StringBuilder();
+                        int firstLine = linesTotal - stackTraceLineCount;
+
+                        // Salvage Javac localized message 'javac.msg.bug' ("An exception has occurred in the
+                        // compiler ... Please file a bug")
+                        if (firstLine > 0) {
+                            final String lineBeforeStackTrace = lines[firstLine - 1];
+                            // One of those two URL substrings should always appear, without regard to JVM locale.
+                            // TODO: Update, if the URL changes, last checked for JDK 21.
+                            if (lineBeforeStackTrace.contains("java.sun.com/webapps/bugreport")
+                                    || lineBeforeStackTrace.contains("bugreport.java.com")) {
+                                firstLine--;
+                            }
+                        }
+
+                        // Note: For message 'javac.msg.proc.annotation.uncaught.exception' ("An annotation processor
+                        // threw an uncaught exception"), there is no locale-independent substring, and the header is
+                        // also multi-line. It was discarded in the removed method 'parseAnnotationProcessorStream',
+                        // and we continue to do so.
+
+                        for (int i = firstLine; i < linesTotal; i++) {
+                            buffer.append(lines[i]).append(EOL);
+                        }
+                        errors.add(new CompilerMessage(buffer.toString(), CompilerMessage.Kind.ERROR));
                     }
                 }
                 return errors;
             }
 
-            // A compiler error occurred, treat everything that follows as part of the error.
-            if (line.startsWith("An exception has occurred in the compiler")) {
-                buffer = new StringBuilder();
-
-                while (line != null) {
-                    buffer.append(line);
-                    buffer.append(EOL);
-                    line = input.readLine();
-                }
-
-                errors.add(new CompilerMessage(buffer.toString(), CompilerMessage.Kind.ERROR));
-                return errors;
-            } else if (line.startsWith("An annotation processor threw an uncaught exception.")) {
-                CompilerMessage annotationProcessingError = parseAnnotationProcessorStream(input);
-                errors.add(annotationProcessingError);
+            if (stackTraceLineCount == 0 && STACK_TRACE_FIRST_LINE.matcher(line).matches()
+                    || STACK_TRACE_OTHER_LINE.matcher(line).matches()) {
+                stackTraceLineCount++;
```

Comment:
`stackTraceLineCount` will only have value `1` or `0`

Reply:
Not true, please check again. Sorry for force-pushing in the meantime, I improved a few things. But with regard to the count, nothing has changed.

Reply:
It is only one place where we assign/increment `stackTraceLineCount` (which I see) and in condition to this block of code we have `stackTraceLineCount == 0`

Reply:
Well, `stackTraceLineCount++` increments the value each time it finds another stack trace element. So for a stack trace of 7 lines, the value will end up to be 7. You said, the value can only be 0 or 1. Feel free to clone the PR and step through it, if the logic seems to complex. Actually, it is pretty simple.

Reply:
> in condition to this block of code we have `stackTraceLineCount == 0`

No, we have 

```java
// If count is 0, match first line
stackTraceLineCount == 0 && STACK_TRACE_FIRST_LINE.matcher(line).matches()
  ||
// Otherwise, match subsequent (other) lines
STACK_TRACE_OTHER_LINE.matcher(line).matches()
```

Maybe with my additional line break and comments, it is clearer for you.

Reply:
Oh, now I get it. I think, you disregarded the fact the `A && B || C` is not the same as `A && (B || C)`. It is, in fact, `(A && B) || C`. Normal logical operator preference, AND binds more strongly then OR.

Reply:
I see 😄 ... I need more ☕